### PR TITLE
RetryIf might need more information and greater control over the execution flow.

### PR DESCRIPTION
**Pull request by newacorn**
Another implementation of RetryIfErr

    * Support for delayed retry logic
    * Support for resetting the request timeout before retrying
    * Users can decide whether to retry based on the request error and the number of retries.
    Estimating the time required for a request to complete is relatively straightforward, but when retries are factored in, it becomes challenging to gauge this time accurately.

At the same time, comment out the logic for retrying non-idempotent requests：
```go
	if !retry {
			break
			// Retry non-idempotent requests if the server closes
			// the connection before sending the response.
			//
			// This case is possible if the server closes the idle
			// keep-alive connection on timeout.
			//
			// Apache and nginx usually do this.
			//
			// This assertion cannot ensure that the server has not read our request.
			// Even if part of the request has been read, retrying for a non-idempotent
			// request is still unsafe. It is better to leave this decision to the user.
			//
			// if err != io.EOF {
			//	break
			// }
		}
```
This logic only covers one case where the error is not io.EOF; in fact, there are many other situations and edge cases. For non-idempotent requests, we should not make assumptions for the user.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -659,9 +669,18 @@ type DialFuncWithTimeout func(addr string, timeout time.Duration) (net.Conn, err
...
+// whether to retry later, or whether to reset the request timeout—should be determined
+// based on the return value of this interface.
+//
+// retryCount indicates which retry attempt this is, with retry counting starting from 1.
+// The first request execution is not considered a retry.
+// err represents the error encountered during the previous request.
+// sleepDuration if is greater than 0, the retry will be executed after this interval.
+// resetReqTime indicates whether to reset the request timeout to its initial value before executing this retry.
+// retry indicates whether to perform this retry. If false, the request function will return immediately.
+type RetryIfErrFunc func(request *Request, retryCount int, err error) (sleepDuration time.Duration, resetReqTime bool, retry bool)
```
Comment (erikdubbelboer) [inline client.go:683]: What is the reason for returning the `sleepDuration`? The user of this function can more easily just call `time.Sleep()` in the `RetryIfErrFunc` itself.

File: client.go [hunk 2]
```diff
@@ -1329,31 +1357,40 @@ func (c *HostClient) Do(req *Request, resp *Response) error {
...
-				break
-			}
+			//
+			// This assertion cannot ensure that the server has not read our request.
+			// Even if part of the request has been read, retrying for a non-idempotent
+			// request is still unsafe. It is better to leave this decision to the user.
+			//
+			// if err != io.EOF {
+			//	break
+			// }
```
Comment (erikdubbelboer) [inline client.go:1387]: I agree, lets just remove the comment to make the code cleaner.

File: client.go [hunk 3]
```diff
@@ -659,9 +669,18 @@ type DialFuncWithTimeout func(addr string, timeout time.Duration) (net.Conn, err
...
-// RetryIfErrFunc defines the signature of the retry if error function.
-// request and error passed to RetryIfErrFunc, if there are any request errors.
-type RetryIfErrFunc func(request *Request, err error) bool
+// RetryIfErrFunc defines an interface used for implementing the following functionality:
+// When the client encounters an error during a request, the behavior—whether to retry,
+// whether to retry later, or whether to reset the request timeout—should be determined
+// based on the return value of this interface.
+//
+// retryCount indicates which retry attempt this is, with retry counting starting from 1.
+// The first request execution is not considered a retry.
```
Comment (erikdubbelboer) [inline client.go:678]: Maybe it's easier to call this `attempts`?

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: client.go (hunk 2 above)
Comment (newacorn) [reply client.go:1387]: ok

File: client.go (hunk 3 above)
Comment (newacorn) [reply client.go]: Makes sense.

**Comment by erikdubbelboer** [comment]
Thanks!

**Pull request merged**